cylc/flow/task_pool.py

```diff
@@ -1619,10 +1644,19 @@ def spawn_task(
         if not self.can_be_spawned(name, point):
             return None
 
-        submit_num, prev_status, prev_flow_wait = (
+        never_spawned, submit_num, prev_status, prev_flow_wait = (
             self._get_task_history(name, point, flow_nums)
         )
 
+        if (
+            not never_spawned and
+            not prev_flow_wait and
+            submit_num == 0
+        ):
+            # Previous instance removed before completing any outputs.
+            LOG.info(f"Not spawning {point}/{name}: already used in this flow")
```

Comment:
Indeed, but how to describe what's happening concisely? 

Basically, the flow (on this branch of the graph) came to a halt at this point because the task was removed earlier by suicide trigger or `cylc remove`. That definitely warrants some kind of log message.

I'll change it to this (subject to further litigation, of course):
```suggestion
            LOG.info(f"Flow stopping at {point}/{name} - task previously removed")
```

Reply:
I don't like "flow blocked" I'm afraid! How about:
```
Skipping {point}/{name} {stringify_flow_nums(flow_nums)} - task previously removed
```

Reply:
I think "flow blocked" is better than your two suggestions, because it correctly indicates that the flow will not continue here because the task was removed.

"Skipping" suggests the flow will skip over the remove task, which is not the case.

"Spawning" (as you'll have seen in the chat) is I think being expunged from user terminology.

Reply:
Anther terminology debate :facepalm:. FYI, I don't think that any of these suggestions will make a lick of sense to users.

"Flow Blocked" is incorrect, the flow isn't "blocked", it's most likely continuing elsewhere in the graph, it just isn't proceeding through this particular task by design.

"Skipping" is incorrect as the flow won't "skip" over the task.

"Not Spawning" is correct (or not *re*-spawning if we're being picky). Note, the term "spawn" is not being "expunged", it was never a user facing term in the first place.

This log message isn't really relevant to users, the removal / suicide was already logged in a way that they can understand, this is an engineers message for debugging the downstream consequences of removal, it is only relevant if the removal mechanism doesn't work for some reason. So we can call it whatever we like and stick it at debug level.

Reply:
>  FYI, I don't think that any of these suggestions will make a lick of sense to users.

I agree, mostly, but unfortunately that doesn't absolve us from coming up with consistent terminology for these behaviours, and documenting it so that users *can* understand it if they want to or need to.

> "Flow Blocked" is incorrect, the flow isn't "blocked", it's most likely continuing elsewhere in the graph, it just isn't proceeding through this particular task by design.

It is correct in the sense that I meant it: the flow is literally blocked *on that branch of the graph*, it doesn't mean the entire flow is blocked or stopped.

> "Skipping" is incorrect as the flow won't "skip" over the task.

Agreed.

> Not Spawning" is correct (or not re-spawning if we're being picky).

Agreed, but  you've been arguing against using "spawn" as a user-facing term anymore on Element chat, and the scheduler log IS user-facing. Users often need to look at the look to figure out what their workflow did, or how it ended up in a particular state.

>  Note, the term "spawn" is not being "expunged", it was never a user facing term in the first place.

Ah, it was, and still is. e.g.:
 - the long-standing pre-Cylc 8 config, `spawn to max active cycle points`
 - in the (user-facing!) scheduler log
 - and in the user guide, where it describes how the workflow advances
 
The word "spawn" is used a bunch of times in the user guide "concurrent flows" section for instance.

In fact, reading that again made me realize that "spawn" is that right user-facing term for that concept. E.g (User Guide) *This triggers a task with no flow numbers. It will not spawn children, and other flows that come by will re-run it.*

> This log message isn't really relevant to users, the removal / suicide was already logged in a way that they can understand, this is an engineers message for debugging the downstream consequences of removal, it is only relevant if the removal mechanism doesn't work for some reason. So we can call it whatever we like and stick it at debug level.
 
 I disagree with this. The removal could have been done, and logged, quite some time ago, and it could have been done automatically by suicide trigger or by scripted use of `cylc remove`. We need to log again, at INFO level, why the flow (on this branch of the graph) is stopping at this point in the graph.

Reply:
OK, I think options are as follows (or variants thereof):
- *flow blocked here, the task was removed earlier*  (note addition of "here" to avoid implicating the entire flow)
- *not spawning task, it was removed earlier*

Given that @oliver-sanders agrees "spawning" is correct, and that (as I've just shown) spawning probably should *remain* a user-facing term, perhaps we should stick with that.

However, it needs to be logged at INFO level to explain why the flow suddenly halts there (on this branch of the graph).

Reply:
> Note, the term "spawn" ... was never a user facing term in the first place.
>
>> Ah, really?!

That's Cylc 7!

I'm very confused by these comments. It was your own argument that users should not need to understand the management (insertion or removal) of task proxies or task pool logic in general which I got on board with. It was a primary goal of SoD to remove the need for us to communicate this internal implementation detail, and a necessary side effect as SoD would make it harder for users to monitor the task pool itself. That's why spawning was retired right, `s/spawn/set-outputs`, users shouldn't need to understand spawning, because SoD will ensure the task proxy is always there when needed right? E.G. https://github.com/cylc/cylc-flow/issues/2143. I agreed with you on this at the time, it was a good idea!

I appreciate this is going back a few years now, maybe reality crept in and the thinking has changed. Perhaps it was a bit ambitions to presume we could hide the internal logic. Maybe we should write up the SoD model and communicate it to users as essential reading?


> It needs to be logged at INFO level though, to explain why the flow does not continue here, at this time.

I'm not convinced it does. We are trying to communicate to users the graph model, not the internal SoD implementation (or at least I thought we were).

Take this graph for example:

```
a => b => c => d
a => x
b => x
c => x
d => x
```

If the user runs `cylc remove //x`, they will get a message along the lines of "x removed".

They don't need to see four "not re-spawning x" messages in a row as the graph proceeds, there is no new information in these messages. Moreover, it isn't communicating anything to them about the graph. Of course x isn't be re-spawned, I removed it! On purpose!

Reply:
My approval still stands of course, I'll leave you two to find something you can agree on.

I was trying to overt a terminology argument not to stoke one, I was ok with the original suggestion.

Reply:
I just think "flow blocked" sounds too negative and some users will see it and think something has gone wrong

Reply:
OK @MetRonnie, let's try to agree. I don't see "flow blocked" as negative, but perhaps we could add more information.

We've established that "skipping" isn't right. Are you happy with either of these, or do you have another suggestion?:

- "flow blocked by previous removal of <task>"
- "not spawning <task> (previously removed)"

I like "flow blocked" because it automatically conveys the fact that the flow will not continue from this point.